lib/google/gax/operation.rb

```diff
@@ -0,0 +1,167 @@
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+require 'time'
+
+# This must be loaded separate from google/gax to avoid circular dependency.
+require 'google/gax/constants'
+require 'google/gax/settings'
+require 'google/longrunning/operations_api'
+require 'google/protobuf/well_known_types'
+
+module Google
+  module Gax
+    # A class used to wrap Google::Longrunning::Operation objects.
+    #
+    # @example General usage example
+    #   require 'google/gax/operation'
+    #
+    #   op = Google::Gax::Operation.new api.methodThatReturnsOperation(name)
+    #   op.done?
+    #   op.reload!
+    #   op.done?
+    #   results = op.results
+    #   raise results if op.error?
+    #
+    # @attribute [r] grpc_op
+    #   @return [Google::Longrunning::Operation] The wrapped grpc
+    #     operation object.
+    # @attribute [r] client
+    #   @return [Google::Longrunning::OperationsApi] The client that handles the
+    #     grpc operations.
+    class Operation
+      attr_reader :grpc_op, :client
+
+      # @param grpc_op [Google::Longrunning::Operation]
+      #   The inital longrunning operation.
+      # @param client [Google::Longrunning::OperationsApi]
+      #   The client that handles the grpc operations.
+      def initialize(grpc_op, client: nil)
+        @grpc_op = grpc_op
+        @client = client ? client : Google::Longrunning::OperationsApi.new
+        @callbacks = []
+      end
+
+      # If the operation is done, returns the result, otherwise returns nil.
+      # If a Class is provided, an instance of that class will try to be
+      # unpacked from the response. If the response is not of the type provided
+      # nil will be returned.
+      #
+      # @param [Class] The class type to be unpacked from the response.
+      # @return [nil | Google::Rpc::Status | Object | Google::Protobuf::Any ]
+      #   The result of the operation
+      def results(response_type: nil)
+        return nil unless done?
+        return @grpc_op.error if error?
+        return @grpc_op.response.unpack(response_type) if response_type
+        @grpc_op.response
+      end
```

Comment:
After thinking about how responses can be unpacked in dynamic languages [here](https://github.com/googleapis/gax-nodejs/pull/64#discussion_r86889911). It may be good to make this response unpacked based on the type returned from the response. I think it can be implemented as follows:

```rb
def results (response_type: nil)
  return nil unless done?
  return @grpc_op.error if error?
  return @grpc_op.response.unpack(response_type) if response_type
  begin
    ResponseType = Google::Protobuf::DescriptorPool.generated_pool.lookup(@grpc_op.type_name).msgclass
    return @grpc_op.response.unpack(ResponseType)
  rescue
    puts "Operation response type not loaded. Unable to unpack."
  end
  @grpc_op.response
end
```

@jmuk thoughts?